documentation/design/user-stories/shared-infra.adoc

```diff
@@ -0,0 +1,277 @@
+
+```
+apiVersion: enmasse.io/v1beta2
+kind: MessagingAddress
+metadata:
+  name: mydlq
+  namespace: myapp
+spec:
+  systemQueue: {}
+---
+apiVersion: enmasse.io/v1beta2
+kind: MessagingAddress
+metadata:
+  name: myexpiry
+   namespace: myapp
+spec:
+  systemQueue: {}
+```
+
+Then create
+
+```
+apiVersion: enmasse.io/v1beta2
+kind: MessagingAddress
+metadata:
+  name: myqueue
+  namespace: myapp
+spec:
+  queue:
+    # Note: a queue 'mydlq' must exist!
+    DeadLetterQueue: mydlq
+    # Note: a queue 'myexpiry' must exist!
+    ExpiryQueue: myexpiry
+
+    # Forwarders require a MessagingConnector to exist, in this case one that is named 'remote1'.
+    forwarders:
+    - name: fwd1
+      remoteAddress: remote1/queue1
+      direction: in
+```
+
+NOTE: the properties that are only applicable to a few types, are specified under the type object as shown above.
+
+=== CRUDLW a messaging endpoint in a namespace where messaging is enabled
+
+Create a `MessagingEndpoint` resource containing either `cluster`, `route`, `ingress`, `loadbalancer` or `nodeport` objects
+that specify the properties. The protocols supported are AMQP, AMQP+TLS,
+AMQP-WS and AMQP-WS+TLS.
+
+The default protocols depends on the endpoint type. For "Ingress" and "Route", the default
+protocols is AMQPS. For NodePort, LoadBalancer and Cluster, the default protocols are AMQP, AMQPS.
+
+==== Configure an in-cluster endpoint with TLS
+
+```
+apiVersion: enmasse.io/v1beta2
+kind: MessagingEndpoint
+metadata:
+  name: endpoint1
+  namespace: myapp
+spec:
+  # TLS configuration for endpoint
+  tls: {}
+
+  protocols:
+    - AMQPS
+    - AMQP
+
+  cluster: {}
+```
+
+==== Configure an endpoint exposed as a route
+
+```
+apiVersion: enmasse.io/v1beta2
+kind: MessagingEndpoint
+metadata:
+  name: endpoint1
+  namespace: myapp
+spec:
+  route:
+    hostname: messaging.example.com # Optional
+```
+
+==== Configure an endpoint exposed as a loadbalancer
+
+```
+apiVersion: enmasse.io/v1beta2
+kind: MessagingEndpoint
+metadata:
+  name: endpoint1
+  namespace: myapp
+spec:
+  tls: {}
+  loadbalancer
+    annotations: {}
+    requireTls: true # Setting this to 'true' will disable non-TLS port
+```
+
+=== CRUDLW a messaging connector in a namespace where messaging is enabled
+
+=== Managing certificates for a messaging endpoint
+
+```
+apiVersion: enmasse.io/v1beta2
+kind: MessagingEndpoint
+metadata:
+  name: endpoint1
+  namespace: myapp
+spec:
+  ...
+
+  # Required for endpoints of type 'route' and 'ingress', as these only support TLS
+  # For other types, specifying `tls` will ensure they expose traffic on TLS
+  tls:
+    # Provide certificate values externally
+    external:
+      # Can be input directly or using a secret
+      certificate:
+        value: PEM-encoded cert
+      key:
+        valueFromSecret:
+          name: "mysecret" # Name of secret containing certificate
+          key: "tls.key"
+    # Self-signed
+    selfsigned: {}
```

Comment:
in the case where many endpoints take the self signed option, would there be.a single CA shared by all the endpoints (like we have now)?  How would a user get the CA certificate?

Reply:
Thats a good point. If we want to keep self-signed (which I'm open to discuss), we could create the initial CA when tenant is created and then use that CA to sign any endpoints that would need creation. It would mean we generate a potentially unused CA, but I feel like it would be cleaner than to auto-create CA when endpoint is created etc (running into the same lifecycle issues as for auto-creating MessagingTenant).

Reply:
I'd lean towards keeping self signed.   I think it makes the getting started experience easier.  I like the of creating the initialCA when the tenant is created. The overhead of it is sufficiently small to not be of a concern.  I suppose the tls section of each endpoint that refers to the self-signed would expose a ca field.

Reply:
Added to the status section example.